Fix `timestamp()` helper step calculation

distantnative opened the pull request:
## Describe the PR
<!--
A clear and concise description of the bug the PR fixes or the feature the PR introduces.
Use this section for review hints and explanations for easier code understanding if necessary.
-->

Looking at #3642, I realized that the `timestamp()` helper at the moment has quite a lot of flaws/non-functional cases.

The helper now properly calculates the nearest step corresponding to the step's unit and choose sensible reference points to do so (e.g. current century for years).

### TODO
- [x] phpmd issue blocking merging https://github.com/phpmd/phpmd/issues/910 (@lukasbestle)
- [ ] Ensure that JS implementation has no conflict with this change
- [ ] test case: 31.12.2021 23:59 rounding by 1 month (overflow into new year)

Thx to @nilshoerrmann for the ideas and input.


## Release notes
<!--
A concise list of changes to be used in the version release notes.
Please use sub-headings for "Features", "Enhancements", "Fixes"...
-->

### Fixes
- `timestamp()` helper now rounds correctly to the nearest step 
- `timestamp()` helper allows passing a custom `base` reference for step rounding


### Breaking changes
<!--
If PR creates known breaking changes, please list them.
If there are no breaking changes, please write "None".
-->

_None_

## Related issues/ideas
<!--
PR relates to issues in the `kirby` repo or ideas on `feedback.getkirby.com`:
-->

- Fixes #3642

## Ready?
<!--
If you feel like you can help to check off the following tasks,
that'd be great. If not, don't worry - we will take care of it.
-->

- [x] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [ ] CI checks pass

<!--
CI runs automatically when the PR is created. You can also run `composer ci` locally.
Running locally requires PHPUnit, PHP-CS-Fixer, Psalm, PHPCPD and PHPMD.
-->

## When merging
<!-- We will take care of the following TODOs when reviewing and merging the PR. -->

- [ ] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)
- [ ] Add changes to release notes draft in Notion

Commit:
Fix `timestamp()` helper step calculation

Fixes #3642

afbora commented on config/helpers.php:
```diff
@@ -800,50 +800,64 @@ function timestamp(?string $date = null, $step = null): ?int
+            break;
+        case 'month':
+            if ((int)date('d', $date) >= 15) {
+                $date  = strtotime('+20 day', $date);
```
Could use `next month` instead adding day

distantnative replied:
Sadly not. Next month seems to add a fixed numbers of dates. E.g. if the date is 2020-08-31, it jumped to October not September when using next month.

afbora replied:
Ohh, I didn't know this bug. So your solution solves this problem with a single line of code 👍

afbora commented on config/helpers.php:
```diff
@@ -800,50 +800,64 @@ function timestamp(?string $date = null, $step = null): ?int
+            break;
+        case 'year':
+            if ((int)date('m', $date) >= 7) {
+                $date  = strtotime('+7 month', $date);
```
Could use `next year` instead adding month

distantnative replied:
Most times true in this case. I'd assume that would only fail in a leap year on the 366 day... it's a tiny edge case but maybe better to stick to the current (though less elegant) solution?

afbora replied:
I didn't fully understand the edge cases. 2020 leap year and my tests:

- `next year` for `2020-02-29 23:59:59` output is `2021-03-01 23:59:59`
- `next year` for `2020-12-31 23:59:59` output is `2021-12-31 23:59:59`
- `next year` for `2020-01-01 00:00:00` output is `2021-01-01 00:00:00`

afbora commented on config/helpers.php:
```diff
@@ -800,50 +800,64 @@ function timestamp(?string $date = null, $step = null): ?int
+            $date = strtotime(date('Y-m-01 00:00:00', $date));
+            break;
+        case 'year':
+            if ((int)date('m', $date) >= 7) {
```
Could use `idate('n', $date)` instead casting.

afbora commented on config/helpers.php:
```diff
@@ -800,50 +800,64 @@ function timestamp(?string $date = null, $step = null): ?int
+            $date = strtotime(date('Y-m-d 00:00:00', $date));
+            break;
+        case 'month':
+            if ((int)date('d', $date) >= 15) {
```
Could use `idate('j', $date)` instead casting

afbora commented on config/helpers.php:
```diff
@@ -790,60 +790,87 @@ function tc(string $key, int $count)
-
-    // on error, convert `false` into `null`
-    return $timestamp ?? null;
+    return null;
```
In what case does this line work? 🤯 😅

nilshoerrmann replied:
This return statement is the fallback if the period `$options` are empty if I understand it correctly.

distantnative replied:
I felt like there should be an ultimate return in case nothing before matched. But I agree that I with the current setup, I think this line can't ever really been reached. 🤔

nilshoerrmann commented on config/helpers.php:
```diff
@@ -790,60 +790,87 @@ function tc(string $key, int $count)
     }
 
+    // convert date to timestamp
     $date = strtotime($date);
```
What happens if `$date` cannot be interpreted by `strtotime`?

afbora replied:
A few lines up `V::date($date)` takes care of it.

afbora replied:
@nilshoerrmann You may be right 🤔 

````php
$date = 'january';

var_dump(strtotime($date));
var_dump(date_parse($date));
````

**Output**

````
int(1611183600)

array(12) {
  ["year"]=>
  bool(false)
  ["month"]=>
  int(1)
  ["day"]=>
  bool(false)
  ["hour"]=>
  bool(false)
  ["minute"]=>
  bool(false)
  ["second"]=>
  bool(false)
  ["fraction"]=>
  bool(false)
  ["warning_count"]=>
  int(1)
  ["warnings"]=>
  array(1) {
    [8]=>
    string(27) "The parsed date was invalid"
  }
  ["error_count"]=>
  int(0)
  ["errors"]=>
  array(0) {
  }
  ["is_localtime"]=>
  bool(false)
}
````

distantnative replied:
@afbora but that looks correct - we wouldn't want to allow just `january` as input and `V::date()` will not allow it - or am I reading this wrong?

afbora replied:
You read it right, I thought wrong 🙂 The both return different results, but `V::date()` works first.

Commit:
`timestamp()`: allow passing base for step

lukasbestle committed:
CI: Disable PHPMD temporarily

False-positive for missing imports in namespace-less files; introduced by changes in https://github.com/getkirby/kirby/pull/3658.
